{% extends "base.html" %}
{% block title %}Expression{% endblock %}
{% block page %}Expression estimation{% endblock %}
{% block mainarea %}
<div id="description">
    <p>
    The FPKM, TPM, and read counts of genes and transcripts inferred by RSEM [1].</p>
    <p>
    Methods: RSEM uses input transcriptome alignment by STAR and estimate expression values using the alignments.
    Expression levels are quantified through FPKM (Fragments Per Kilobase of transcript per Million mapped reads),
    TPM, and read counts.
    </p>
</div>
<hr>
    {{ expr_html }}
    <div margin-left="2em">* Gene expression tables are in the folder.</div>
<hr>

<div>
    <p>
    'TPM' stands for Transcripts Per Million. It is a relative measure of transcript abundance. The sum of all
    transcripts' TPM is 1 million. 'FPKM' stands for Fragments Per Kilobase of transcript per Million mapped
    reads. It is another relative measure of transcript abundance. If we define l_bar be the mean transcript
    length in a sample, which can be calculated as
    </p>
    <p>l_bar = \sum_i TPM_i / 10^6 * effective_length_i (i goes through every transcript),</p>

    <p>
    the following equation is hold:
    </p>
    <p>FPKM_i = 10^3 / l_bar * TPM_i.</p>
    <p>
    'count' is the sum of the posterior probability of each read comes from this transcript over all reads.
    Because 1) each read aligning to this transcript has a probability of being generated from background noise;
    2) RSEM may filter some alignable low quality reads, the sum of expected counts for all transcript are generally
    less than the total number of reads aligned.
    </p>
    <p>
    We can see that the sum of FPKM is not a constant across samples.
    </p>
</div>
<p>[1]: Li B, Dewey CN. RSEM:accurate transcript quantification from RNA-Seq data with or without a reference genome. BMC Bioinformatics. 2011; 12:323 </p>

{% endblock %}