{% extends "base.html" %}
{% block title %}Alignment QC{% endblock %}
{% block page %}Alignment QC{% endblock %}
{% block sidebar %}
  <li><a href='workflow.html'>Workflow</a></li>
  <li><a href='project.html'>Project summary</a></li>
  <li><a href='readsdata.html'>Data</a></li>
  <li><a href='alignment_stat.html'>Alignments</a></li>
  <li class="active"><a href='alignment_QC.html'>Alignment QC</a></li>
  <li><a href='expression.html'>Expression estimation</a></li>
  <li><a href='diffexpr.html'>Differential expression</a></li>
  <li><a href='genesets.html'>Pathway analysis</a></li>
  <li><a href='versions.html'>Software</a></li>
{% endblock %}
{% block mainarea %}
  <h1 class="page-header" style="text-align:center">mRNAseq QC</h1>
  <div class="table-responsive">
    {{bam_qc_html}}
  </div>
    <dl>
      <dt>Gene Body Coverage</dt>
      <dd>The plot shows the relative abundance of sequencing read across the gene length from 5' to 3'.
          It is a good indicator that whether reads have bias toward one or both ends, which could be
          caused by RNA degradation.</dd>
      <dt>Junction Saturation</dt>
      <dd>The curves show the diversity of annotated and noval splic junctions. The former reveals if there
          are diverse RNA population in the sequencing sample, thus if the sample has enough depth. The later
          indicates how many new, un-annotated junctions are revealed by increasing number of sequencing reads.</dd>
        <dt>Splicing Junction</dt>
        <dd>The pie chart shows the percentage of annotated and novel junctions presented in the sample.</dd>
        <dt>ReadPair Inner Distance</dt>
        <dd>The histogram shows the distribution of inner distance (between two closest ends) of the reads
            pairs estimated from the alignment. The accurate inner distance cannot be estimated from the
            single end sequencing data</dd>
        <dt>Read Dupliction</dt>
        <dd>The plot shows the read duplication level and theoretical duplication level</dd>
    </dl>

{% endblock %}
