{% extends "base.html" %}
{% block title %}workflow{% endblock %}
{% block page %}Workflow{% endblock %}
{% block sidebar %}
  <li class="active"><a href='workflow.html'>Workflow</a></li>
  <li><a href='project.html'>Project summary</a></li>
  <li><a href='readsdata.html'>Data</a></li>
  <li><a href='alignment_stat.html'>Alignments</a></li>
  <li><a href='alignment_QC.html'>Alignment QC</a></li>
  <li><a href='expression.html'>Expression estimation</a></li>
  <li><a href='diffexpr.html'>Differential expression</a></li>
  <li><a href='genesets.html'>Pathway analysis</a></li>
  <li><a href='versions.html'>Software</a></li>
{% endblock %}
{% block mainarea %}
            <article class="result-content">
                <h1 class="page-header" style="text-align:center">Workflow</h1>
                <p>
                    Sequencing reads were trimmed to remove sequencing adaptors and low quality bases. And subsequently aligned
                    to the annotated reference genome using aligner STAR [1]. Quantification of expression for each gene is calculated
                    by RSEM[2], which produce both Transcripts Per Million reads (TPM), and expected read count. The expected read counts
                    are used for differential gene expression analysis using EdgeR [3] or EBSeq [4]
                </p>
                <img src="workflow.png" alt="[Workflow Chart]">
            </article>
                <footer>[1] Dobin A <i>et al</i> (2013) STAR: ultrafast universal RNA-seq aligner. Bioinformatics 29:15-21.
                </footer>
                <footer>[2] Li B, Dewey CN ( 2011) RSEM: accurate transcript quantification from RNA-Seq data with or without a reference genome.
                    BMC Bioinformatics, 12:323
                </footer>
                <footer>[3] Robinson MD, McCarthy DJ, Smyth GK (2010). edgeR: a Bioconductor package for
                    differential expression analysis of digital gene expression data. Bioinformatics 26, 139â€“140.
                </footer>
                <footer>[4] Leng N <i>et al</i> (2013) EBSeq: an empirical Bayes hierarchical model for inference in RNA-seq experiments.
                    Bioinformatics 29:1035-1043.
                </footer>
{% endblock %}
